docs/adr/005-publishing-api-debounce.md

```diff
@@ -0,0 +1,123 @@
+
+For historical reasons in how upstream CMSs work internally, a single (ostensibly atomic) change to
+a piece of content may actually be represented as a series of changes in Publishing API, in turn
+triggering several messages in rapid succession to downstream consumers. For example, updating
+"links" between different content items involves a separate operation from updating the content of a
+page itself.
+
+Currently, this presents two problems for Search API v2:
+- Discovery Engine has a strict rate limit on how many times a document can be updated within the
+  space of a second (roughly once), which we exceed due to processing several messages
+  simultaneously across workers
+- Having a high number of workers listening to the message queue leads to potential out-of-order
+  processing of messages, meaning that the last update performed upstream isn't necessarily the last
+  one to be processed by us
+
+Both of these points have the potential to lead to data integrity issues, with newer content in the
+Discovery Engine datastore being overwritten by outdated content (out of order), or newer content
+not being written at all (hitting rate limits).
+
+Neither of these problems were encountered with the previous Elasticsearch-based Search API v1, as
+it did not have rate limiting, and made use of the messages' payload version and Elasticsearch's
+versioning ability to avoid overwriting newer documents with older ones.
+
+## Considered options
+
+Discovery Engine does not offer transactions, document versioning, or atomic read-and-update-if
+style operations, which limits our options to:
+
+### Do nothing
+
+Currently, any message coming through from Publishing API triggers an update of the document
+directly in Discovery Engine (regardless of the document version or any other workers concurrently
+dealing with the same document).
+
+This is not a long-term option, because:
+- Having any number of missing or stale documents in search over an extended period of time is
+  unacceptable
+- The amount of errors is polluting our error management tool (and muting/ignoring them may mask
+  future unrelated bugs)
+
+### Reduce worker count to 1 to enfore in-order processing
+
+In theory, limiting incoming content update processing to a single worker in Search API v2 could
+bypass this problem, as under those circumstances RabbitMQ will guarantee in-order delivery and the
+Discovery Engine API calls would have enough breathing room to not run into rate limiting.
+
+However, this isn't acceptable as:
+- we would lose redundancy in case of worker/Kubernetes node failure
+- a single worker would struggle to keep up with peak load (or full re-sync scenarios) leading to
+  exceeding our desired timeframe of how long it takes for an updated document to be visible in
+  search (<1 minute)
+
+### Implement a locking mechanism
+
+A relatively basic locking mechanism using an AWS ElastiCache Redis instance as part of the
```

Comment:
It's been mentioned in slack, but could you update here too that you'll be using the shared [backend-redis](https://docs.publishing.service.gov.uk/repos/govuk-aws/architecture/decisions/0029-combine-api-redis-into-backend-redis.html) instance?